racket/collects/file/private/glob.rkt

```diff
@@ -388,6 +399,48 @@
             (string c)])))
       (string-join str* ""))))
 
+(define GLOB-WILDCARD-CHAR* '(#\* #\? #\[ #\] #\{ #\} #\, #\\))
+
+(define (glob-quote/string str)
+  (define str*
+    ;; add #\\ before all wildcards
+    (for/list ([c (in-string str)])
+      (if (memq c GLOB-WILDCARD-CHAR*)
+        (string #\\ c)
+        (string c))))
+  (apply string-append str*))
+
+(module+ test
+  (test-case "glob-quote/string"
+    (check-equal? (glob-quote/string "a") "a")
+    (check-equal? (glob-quote/string "a*") "a\\*")
+    (check-equal? (glob-quote/string "*][?") "\\*\\]\\[\\?")
+    (check-equal? (glob-quote/string "racket/**/base") "racket/\\*\\*/base")
+    (check-equal? (glob-quote/string "},{foo,bar}") "\\}\\,\\{foo\\,bar\\}")
+    (check-equal? (glob-quote/string "\\") "\\\\")))
+
+(define (glob-unquote/string str)
+  (define str*
+    ;; examine `str` in reverse, remove #\\ from escaped wildcards
+    (let loop ([c* (reverse (string->list str))]
+               [i (- (string-length str) 1)])
+      (cond
+        [(null? c*)
+         '()]
+        [(and (memq (car c*) GLOB-WILDCARD-CHAR*)
+              (escaped? str i))
+         (cons (string (car c*)) (loop (cddr c*) (- i 2)))]
+        [else
+         (cons (string (car c*)) (loop (cdr c*) (- i 1)))])))
+  (apply string-append (reverse str*)))
+
+(module+ test
+  (test-case "glob-unquote/string"
+    (check-equal? (glob-unquote/string "a") "a")
+    (check-equal? (glob-unquote/string "foo\\*rkt") "foo*rkt")
+    (check-equal? (glob-unquote/string "?\\?\\]\\[\\*") "??][*")
+    (check-equal? (glob-unquote/string "\\}a\\,") "}a,")))
```

Comment:
Then maybe `glob-unquote` should be renamed to avoid confusion, because I don't think it needs to satisfy those properties.

What I'm thinking is, `glob` needs to do something special for every wildcard in a pattern. If the wildcard is un-escaped, then it needs to match a set of files. Otherwise if the wildcard is escaped, then it needs to put the wildcard character in the string. Right now `glob-unquote` is handling that "Otherwise" case. (And unfortunately the codes for handling escaped & unescaped wildcards are in 2 different places in this module.)

Put another way:
```
glob-quote : pattern -> pattern
glob-unquote : pattern -> filename
```

Reply:
So why should `glob-element->filename` on `\normal` return `\normal`? (written as `"\\normal"`) Why that and not `normal` or an error?

Looking more closely at how it's used, it makes more sense that it's just transforming one element. And in particular it's only used on elements where both `has-**?` and `has-glob-pattern?` have returned false. Should its signature be updated to include that constraint?

If that's the case, would `\normal` or `?mark` be invalid inputs?